tests/test_requests.py

```diff
@@ -493,3 +493,65 @@ async def app(scope, receive, send):
     client = test_client_factory(app)
     response = client.get("/")
     assert response.json() == {"json": "Send channel not available"}
+
+
+def test_request_body_then_request_body(test_client_factory):
+    # If the request body is read, then ensure that instantiating a
+    # request a second time can return the content again.
+    async def app(scope, receive, send):
+        request = Request(scope, receive)
+        body = await request.body()
+        request2 = Request(scope, receive)
+        body2 = await request2.body()
+        response = JSONResponse({"body": body.decode(), "body2": body2.decode()})
+        await response(scope, receive, send)
+
+    client = test_client_factory(app)
+
+    response = client.post("/", data="abc")
+    assert response.json() == {"body": "abc", "body2": "abc"}
+
+
+def test_request_stream_then_request_body(test_client_factory):
+    # If the request has been streamed, then ensure that instantiating a
+    # request a second time raises an exception when attempting to read content.
+    async def app(scope, receive, send):
+        request = Request(scope, receive)
+        chunks = b""
+        async for chunk in request.stream():
+            chunks += chunk
+
+        request2 = Request(scope, receive)
+        try:
+            body = await request2.body()
+        except RuntimeError:
+            body = b"<stream consumed>"
+
+        response = JSONResponse({"body": body.decode(), "stream": chunks.decode()})
+        await response(scope, receive, send)
+
+    client = test_client_factory(app)
+
+    response = client.post("/", data="abc")
+    assert response.json() == {"body": "<stream consumed>", "stream": "abc"}
+
+
+def test_request_body_then_request_stream(test_client_factory):
+    # If the request body is read, then ensure that instantiating a
+    # request a second time can stream the content.
+    async def app(scope, receive, send):
+        request = Request(scope, receive)
+        body = await request.body()
+
+        request2 = Request(scope, receive)
+        chunks = b""
+        async for chunk in request2.stream():
+            chunks += chunk
+
+        response = JSONResponse({"body": body.decode(), "stream": chunks.decode()})
+        await response(scope, receive, send)
+
+    client = test_client_factory(app)
+
+    response = client.post("/", data="abc")
+    assert response.json() == {"body": "abc", "stream": "abc"}
```

Comment:
```suggestion
    assert response.json() == {"body": "abc", "stream": "abc"}


def test_request_body_then_replace_body(test_client_factory):
    import zlib

    data = "Hello, world!"
    gzipped_payload = zlib.compress(data.encode())

    def ungzip_middleware(app):
        async def wrapped_app(scope, receive, send):
            dec = zlib.decompressobj()  # offset 16 to skip the header
            buffer = bytearray()
            async def wrapped_rcv():
                more_body = True
                while more_body:
                    msg = await receive()
                    more_body = msg.get("more_body", False)
                    buffer.extend(msg["body"])
                    if len(buffer) > 16 or not more_body:
                        msg["body"] = dec.decompress(buffer)
                        yield msg
                        buffer.clear()
            await app(scope, wrapped_rcv().__anext__, send)
        return wrapped_app

    def log_request_body_middleware(app):
        async def wrapped_app(scope, receive, send):
            await Request(scope, receive).body()
            await app(scope, receive, send)
        return wrapped_app

    async def app(scope, receive, send):
        request = Request(scope, receive)
        body = await request.body()

        response = JSONResponse({"body": body.decode()})
        await response(scope, receive, send)

    app = log_request_body_middleware(ungzip_middleware(app))

    client = test_client_factory(app)

    response = client.post("/", data=gzipped_payload)
    assert response.json() == {"body": data}

```

@gnat this is what I'm referring to

`ungzip_middleware` is a perfectly valid ASGI middleware. It could even be inside of a mounted ASGI app or something like that. If we implement this PR we completely break this sort of thing and it breaks in a very non-intuitive way. We're essentially dealing with [cache invalidation](https://martinfowler.com/bliki/TwoHardThings.html)